Simplenote/src/main/java/com/automattic/simplenote/NoteEditorActivity.java

```diff
@@ -235,6 +263,114 @@ public void showTabs() {
         mViewPager.setPagingEnabled(true);
     }
 
+    public void setSearchMatchBarVisible(boolean isVisible) {
+        if (mSearchMatchBar != null) {
+            mSearchMatchBar.setVisibility(isVisible ? View.VISIBLE : View.GONE);
+        }
+    }
+
+    private void setUpSearchMatchBar(Intent intent) {
+        if (mSearchMatchIndexes == null) {
+            String matchOffsets = intent.getStringExtra(NoteEditorFragment.ARG_MATCH_OFFSETS);
+            String[] matches = matchOffsets != null ? matchOffsets.split("\\s+") : new String[]{};
+            String[] matchesStart = new String[matches.length / MATCH_INDEX_COUNT];
+
+            // Get "start" item from matches.  The format is four space-separated integers that
+            // represent the location of the match: "column token start length" ex: "1 3 3 7"
+            for (int i = MATCH_INDEX_START, j = 0; i < matches.length; i += MATCH_INDEX_COUNT, j++) {
+                matchesStart[j] = matches[i];
+            }
+
+            // Remove duplicate items with linked hash set and linked list.
```

Comment:
what is the reason we expect duplicates?

Reply:
The string we get from `ARG_MATCH_OFFSETS` and full-text search can have duplicates.  One case is when the title text has a match.  An example might help.  Below is a string we get with the set of matches in the format `column token start length`.  We want to grab the `start` value from those matches to know where to scroll in the note.  You can see the first two sets are not identical, but they have the same `start` value.  That makes it seem like there are seven matches for the search term when there are actually six; the title is counted twice.

```
INDEX  00 01 02 03 | 04 05 06 07 | 08 09 10 11 | 12 13 14 15 | 16 17 18 19 | 20 21 22 23 | 24 25 26 27
MATCH   0  0  0  6 |  1  0  0  6 |  1  0 17  6 |  1  0 25  6 |  1  0 43  6 |  1  0 55  6 |  1  0 89  6
START         0    |        0    |       17    |       25    |       43    |       55    |       89 
```

Reply:
is this due to the `title`/`content` data structure or a regexp we're using?

I don't doubt what you are saying is true, just trying to understand _why_ it happens and maybe hint at what things would be helpful in a comment to the next dev who comes around and naively says "there shouldn't be any duplicates, let's just remove this" 🙂

Reply:
> Do you think we should be explicit and say it's due to title/content and full-text search in a comment?

I've seen too many people come around and "fix" things or "improve" things by messing with stuff that's non-obvious from the code. The comment indicates that we're removing duplicates, which at worst can be inferred by the immediately-following code.

What can't be reasonably inferred is why we expect them in the first place.

Something on the order of…

```java
// due to the way that sqlite interacts with the title and content
// we can get duplicate matches from time to time and therefore
// need to remove them
```

Reply:
I updated the comment in 7b6f67ad.